podresources APIs: concrete resources apis: implement GetAllocatableResources

ffromani opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Implement the GetAllocatableResources endpoint proposed in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2043-pod-resource-concrete-assigments

**Which issue(s) this PR fixes**:
Fixes # N/A

**Special notes for your reviewer**:
This PR depends on the changes implemented in https://github.com/kubernetes/kubernetes/pull/93243/
Replaces https://github.com/kubernetes/kubernetes/pull/94478

The original plan is to also implement the Watch endpoint, also replacing  and https://github.com/kubernetes/kubernetes/pull/94612,  but this needs to be postponed to 1.21 for capacity reasons.
KEP amendment PR: https://github.com/kubernetes/enhancements/pull/2125

**Does this PR introduce a user-facing change?**:
```release-note
Implement the GetAvailableResources in the podresources API.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
- [KEP]:  https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2043-pod-resource-concrete-assigments
```

ffromani committed:
e2e: node: bootstrap podresources tests

Start e2e tests for the existing List() API.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: add tests for GetAllocatableResources

Add e2e tests for the new GetAllocatableResources API.
The tests are added in the `podresources_test` suite
created previously in this series.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/sig node

fromanirh commented:
this PR is now reviewable! please note it depends on the changes made in https://github.com/kubernetes/kubernetes/pull/93243/ (hence https://github.com/kubernetes/kubernetes/pull/95734/commits/7aaef8a9248f8fe3394365986e8d251412bf8c26 is duplicated here)

The Watch() implementation is missing as we are still tuning the implementation and fixing the e2e tests. Please let me know (@RenaudWasTaken and anyone which wants to review!) if is better to add the Watch implementation here, as I plan to do, or better a separate PR. Please note that PR would depend on https://github.com/kubernetes/kubernetes/pull/93243/ *and* on this one, which can make the review awkward - this is the reason why I'm inclined to add the missing changes here.

SergeyKanzhelev commented:
/ok-to-test

fromanirh commented:
> This PR [may require API review](https://git.k8s.io/community/sig-architecture/api-review-process.md#what-apis-need-to-be-reviewed).
> 
> If so, when the changes are ready, [complete the pre-review checklist and request an API review](https://git.k8s.io/community/sig-architecture/api-review-process.md#mechanics).
> 
> Status of requested reviews is tracked in the [API Review project](https://github.com/orgs/kubernetes/projects/13).

After reviewing the docs I don't think anymore this PR changes the API in the same way the label requires. Fixing the description (and trying to remove the label)

fromanirh commented:
/remove-kind

fromanirh commented:
/remove-kind api-change

fromanirh commented:
/retest

AlexeyPerevalov commented on pkg/kubelet/apis/podresources/server.go:
```diff
@@ -79,7 +83,60 @@ func (p *podResourcesServer) List(ctx context.Context, req *v1alpha1.ListPodReso
+				numaNodes = append(numaNodes, numaNode)
+			}
+		}
+		sort.Stable(numaIDs(numaNodes))
```
why do we need a stable sort here, it has more swap operation than usual Sort?

fromanirh replied:
No real reason. Changing to `sort.Sort`.

RenaudWasTaken commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1alpha1/api.proto:
```diff
@@ -39,10 +39,22 @@ message PodResources {
 message ContainerResources {
```
Split this file in its own commit and the pb.go in its own too :)

fromanirh replied:
sure thing

RenaudWasTaken reviewed (commented):
Took a first stab at it (haven't reviewed the tests yet), overall looks good :)

AlexeyPerevalov commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -902,7 +902,14 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 		// Update internal cached podDevices state.
 		m.mutex.Lock()
-		m.podDevices.insert(podUID, contName, resource, allocDevices, resp.ContainerResponses[0])
+		allocDevicesWithNUMA := make(devicePerNUMA)
```
> I might be nitpicking but the map building should probably be in a function :)

ok, I'll do. The PR for this patch was initially https://github.com/kubernetes/kubernetes/pull/93243.
So I'll update that PR, and if @fromanirh gives me access to this, this one too.

fromanirh replied:
@AlexeyPerevalov I'm fine giving your permissions, but in general is fine if you just fix your PR, I can rebase on top of yours and pick up your changes here.
I'm doing this because this PR depends on the data structure (and on some internal functions) changes you do on your PR, hence I replicated your commit here. Of course once your PR gets merged I'll remove my dupe.

AlexeyPerevalov commented on pkg/kubelet/cm/devicemanager/checkpoint/checkpoint.go:
```diff
@@ -34,7 +34,7 @@ type PodDevicesEntry struct {
 	ContainerName string
 	ResourceName  string
-	DeviceIDs     []string
+	DeviceIDs     map[int64][]string
```
from devicemanager, yes it could be here

fromanirh commented:
/retest

fromanirh commented:
the implementation of the Watch() endpoint (+ e2e tests) is coming soon

dims commented:
/unassign

AlexeyPerevalov commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -902,7 +902,14 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
+				allocDevicesWithNUMA[node.ID] = append(allocDevicesWithNUMA[node.ID], dev)
+			}
+		}
+		m.podDevices.insert(podUID, contName, resource, allocDevicesWithNUMA, resp.ContainerResponses[0])
```
I started to do it, but it requires a lot of changes. So maybe it will be in separate PR or definitly in separate patch.

RenaudWasTaken replied:
I think this is resolved in the PR this is based on!

RenaudWasTaken commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1alpha1/api.proto:
```diff
@@ -18,6 +18,15 @@ option (gogoproto.goproto_unrecognized_all) = false;
 // node resources consumed by pods and containers on the node
 service PodResourcesLister {
     rpc List(ListPodResourcesRequest) returns (ListPodResourcesResponse) {}
+    rpc GetAllocatableResources(AllocatableResourcesRequest) returns (AllocatableResourcesResponse) {}
```
Looks like this escaped my gaze when reading the KEP (and I noticed that it's not very well documented in the kep), could you describe a bit how you expect to use this?
Thanks a lot!

fromanirh replied:
Yeah, this is only mentioned too quickly in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2043-pod-resource-concrete-assigments#topology-aware-scheduling. Let me post a clarification to the KEP

fromanirh replied:
https://github.com/kubernetes/enhancements/pull/2115

RenaudWasTaken commented on pkg/kubelet/apis/podresources/server_test.go:
```diff
@@ -171,3 +184,136 @@ func TestListPodResources(t *testing.T) {
 	}
 }
+
+func TestAllocatableResourcess(t *testing.T) {
```
Nit: `TestAllocatableResourcess` -> `TestAllocatableResources`

fromanirh replied:
good point, will fix.

RenaudWasTaken commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1alpha1/api.proto:
```diff
@@ -18,6 +18,15 @@ option (gogoproto.goproto_unrecognized_all) = false;
+
+// AllocatableResourcesResponses contains informations about all the devices known by the kubelet
+message AllocatableResourcesResponse {
+    repeated ContainerDevices devices = 1;
```
The reuse of containerDevices here is kindof odd, I wonder if we should use a different structure name for it, especially if this is going to be used to sort devices by Numa nodes and by resource name

fromanirh replied:
The intention is to provide an output as much compatible as possible to the output of the other API calls. In this regard reusing some internal messages felt OK (same data). Open to suggestions.

RenaudWasTaken replied:
I pretty much expect this to return resources grouped by resourceName and that's it.
The fact that there might be multiple time the same resource name kinds of point to me that we either:
a) Don't want to group them by resource name
b) Want to have a different "explicit" structure, something that says that we have grouped devices by NUMA nodes.

fromanirh replied:
Now I fully understand your point. And that's a good point indeed. Let me think about the best approach.

RenaudWasTaken commented on pkg/kubelet/apis/podresources/server.go:
```diff
@@ -79,7 +83,60 @@ func (p *podResourcesServer) List(ctx context.Context, req *v1alpha1.ListPodReso
+	for resourceName := range allDevices {
+		resourceNames = append(resourceNames, resourceName)
+	}
+	sort.Strings(resourceNames)
```
We might need to specify the order (in the KEP), or if we don't guarantee it, specify in the API description that the order is not guaranteed.

fromanirh replied:
Good point. Both approaches have pros and cons, but I'm inclined towards NOT requiring the ordering in the API contract (thus in the KEP). I will update accordingly.

fromanirh replied:
Documented in https://github.com/kubernetes/enhancements/pull/2115

RenaudWasTaken reviewed (commented):
I haven't reviewed the tests yet :) but we are going in the right direction!
I'm also confident that at our current pace we should make the code freeze :) !

fromanirh commented on test/e2e_node/podresources_test.go:
```diff
@@ -0,0 +1,277 @@
+	}
+}
+
+func scanPodResources(podIdx int, pr *kubeletpodresourcesv1alpha1.PodResources) int {
```
I'd go with `countPodResources`

fromanirh commented on test/e2e_node/podresources_test.go:
```diff
@@ -0,0 +1,277 @@
+			podName: "pod-01",
+		},
+	})
+	podResources, noResources = getPodResources(cli)
```
Good point. Let's retry few times.

RenaudWasTaken replied:
`gomega.Eventually` might be useful here

fromanirh commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1alpha1/api.proto:
```diff
@@ -18,6 +18,15 @@ option (gogoproto.goproto_unrecognized_all) = false;
 // node resources consumed by pods and containers on the node
 service PodResourcesLister {
```
This is what I wanted to do with https://github.com/kubernetes/kubernetes/pull/95734/commits/0c405e80dcbd1bb1e16670ed5f903ca9e63a88f4 - I guess things got a bit messed up in the upload? Will rebase and re-upload shortly so we can start from a clean(er) base

ffromani reviewed (commented):
Thanks @RenaudWasTaken for your reviews! We also want to add the implementation of Watch() endpoint but this is turning out a bit more complex than initially planned. I will update about the status shortly and we will see how we can move forward (e.g. postponing Watch() to 1.21)

AlexeyPerevalov commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -129,19 +139,26 @@ func (s *sourcesReadyStub) AllReady() bool          { return true }
+	var onlineCPUs cpuset.CPUSet
+	var err error
+
+	onlineCPUs, err = topology.OnlineCPUs()
```
you can get cpus from machineInfo
also you don't take into account reservedCPU, as a result it's a capacity but not allocatable. But we need to know which one cpu kubelet will never use for cpu allocation.

fromanirh replied:
Does machineInfo take into account actually online CPUs?
good point about reserved cpus, will handle that

fromanirh replied:
Yes it does: https://github.com/google/cadvisor/blob/9b7851c9fb0efb02e330ec88a8b54569e7a87529/utils/sysfs/sysfs.go#L338

So we can get rid of this code, and I'll take into account the `reserved cpus` like @AlexeyPerevalov suggested/

ffromani committed:
node: podresources: implement GetAllocatableResources API

Extend the podresources API implementing the GetAllocatableResources endpoint,
as specified in the KEPs:

https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2043-pod-resource-concrete-assigments
https://github.com/kubernetes/enhancements/pull/2404

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
```
/usr/local/bin/entrypoint.sh: line 31:    17 Killed                  /usr/local/bin/runner.sh ./test-infra/jenkins/bootstrap.py --job="${JOB_NAME}" --service-account="${GOOGLE_APPLICATION_CREDENTIALS}" --upload="${BOOTSTRAP_UPLOAD_BUCKET_PATH}" "$@"
```
will retry in a little while

fromanirh commented:
/retest

fromanirh commented:
/retest

dims commented:
/uncc

fromanirh commented:
it seems something weird is happening in the node-e2e lane. Let me retry one last time before to check/report the issue.

fromanirh commented:
/retest

SergeyKanzhelev commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -77,6 +77,13 @@ type Manager interface {
 	// and other resource controllers.
 	GetTopologyHints(*v1.Pod, *v1.Container) map[string][]topologymanager.TopologyHint
+
+	// GetCPUs implements the podresources.GetCPUs interface to provide allocated
```
random nit comment:

```suggestion
	// GetCPUs implements the podresources.CPUsProvider interface to provide allocated
```

fromanirh replied:
it's a welcome nit! fixing

fromanirh replied:
this actually is in the parent PR whose commits I copied in this one to make it compile, so I'll fix on a separate commit.

AlexeyPerevalov replied:
yes, this commit from https://github.com/kubernetes/kubernetes/pull/93243, that PR already got lgtm from @RenaudWasTaken. I updated comment there.

SergeyKanzhelev commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -116,6 +123,9 @@ type manager struct {
 	// stateFileDirectory holds the directory where the state file for checkpoints is held.
 	stateFileDirectory string
+
+	// onlineCPUSet is the set of online CPUs as reported by the system
```
```suggestion
	// assignableCPUs is the set of online CPUs as reported by the system
```

SergeyKanzhelev commented on pkg/kubelet/cm/cpumanager/cpu_manager_test.go:
```diff
@@ -639,15 +643,8 @@ func TestCPUManagerGenerate(t *testing.T) {
-						t.Errorf("Expected topology to be non-nil for policy '%v'. Have: %q", rawMgr.policy.Name(), rawMgr.topology)
-					}
+				if rawMgr.topology == nil {
+					t.Errorf("Expected topology to be non-nil for policy '%v'. Have: %q", rawMgr.policy.Name(), rawMgr.topology)
```
It is nil, why format it? Or it meant to be something else?
```suggestion
					t.Errorf("Expected topology to be non-nil for policy '%v'.", rawMgr.policy.Name())
```

fromanirh replied:
thanks, let me simplify this code.

ffromani committed:
node: e2e: docs and fix for teardownSRIOVConfig

Document why teardownSRIOVPod has to wait for all the containers
to be gone before to end, and why is important.

Additionally, change the code to wait for all the containers to be gone,
not just the first. This is both a little cleaner and a little safer,
even though it seems the current code caused no issues so far.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: e2e: run deleteSync in parallel

speedup the cleanup after testcases deleting pods in separate
goroutines.
The post-test cleanup stage must be done carefully since pod require
exclusive allocation - so pods must take all the steps to properly
cleanup the tests to avoid to pollute the environment, but
this has a negative effect on test duration (take longer).

Hence, we add safe speedups like doing pod deletions in parallel.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented on pkg/kubelet/cm/cpumanager/cpu_manager_test.go:
```diff
@@ -639,15 +643,8 @@ func TestCPUManagerGenerate(t *testing.T) {
 				if rawMgr.policy.Name() != testCase.expectedPolicy {
 					t.Errorf("Unexpected policy name. Have: %q wants %q", rawMgr.policy.Name(), testCase.expectedPolicy)
 				}
-				if rawMgr.policy.Name() == string(PolicyNone) {
```
with my changes this is always false... (hence deleted)

AlexeyPerevalov replied:
I agree with such architectural decision in the code itself. But let's think about meaning of the return value of GetAssignableCPUs, it's  CPUs available for exclusive allocation, that's why static_policy  evaluates it and exclude reserved cpus. So we need to decide, what is assignable cpu in case of none policy, because where is no exclusive isolation in this case. So I agree with removed code, that nil topology for none policy is expected topology.

fromanirh replied:
That's a good point, thanks for raising. I see as viable only this approach or report nothing, because if this `GetAssignableCPUs` needs to report cpus available for exclusive allocation, policy None must report empty set (which is by side effect also greatly simplifies the code!). It really depends on clarifying this point and having it specified.

swatisehgal replied:
By definition, assignable CPUs are the ones that can be exclusively allocated to pods that meet the exclusivity requirement (ie guaranteed QoS class and integral CPU request). I see the reason why previously GetAssignableCPUs was not included in the Policy interface as assignability of CPUs as a concept is only applicable in case of static policy i.e. scenarios where workloads CAN get exclusive access to core(s). By default (in case of None policy), kubernetes relies on CFS quota to enforce CPU limits on pod containers and not on cpu affinity for cpu assignment. 
So, passing back an empty set seems fine to me in case of none policy but it feels like saying that both the polices have assignable CPUs (by adding  GetAssignableCPUs() in the policy interface) and then later in case of none policy returning an empty set but I guess we do that in go all the time :)
Similarly for topology field in case of none policy, first we say that in case of none policy we have a topology and then  always return nil. I wonder what is the motivation for this change? I think if we would like to proceed with these changes, we should add comments in case of none policy too (like we have for static policy [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L35)) to capture these for future developers.

AlexeyPerevalov replied:
so from my point of view:
extending policy interface with GetAssignableCPUs - is ok
keeping topology in NonePolicy is unnecessary, since GetAssignableCPUs will always return nil.

fromanirh replied:
Works for me and as nice bonus makes the code simpler :)
I will summarize this conversation in a code comment.

fromanirh commented on pkg/kubelet/cm/cpumanager/cpu_manager_test.go:
```diff
@@ -639,15 +643,8 @@ func TestCPUManagerGenerate(t *testing.T) {
-						t.Errorf("Expected topology to be nil for 'none' policy. Have: %q", rawMgr.topology)
-					}
-				}
-				if rawMgr.policy.Name() != string(PolicyNone) {
```
same here

fromanirh commented:
Since https://github.com/kubernetes/kubernetes/pull/93243 added `/kind api-change`, I'm adding back as well.

fromanirh commented:
/kind api-change

dims commented:
/milestone v1.20

ffromani committed:
node: podresources: add GetAllocatableResources API

Extend the podresources API adding the GetAllocatableResources endpoint,
as specified in the KEP
https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2043-pod-resource-concrete-assigments

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
It seems I'm also hitting https://github.com/kubernetes/test-infra/issues/19910 ?

fromanirh commented:
/retest
the failure on the bazel-test lane seems unrelated

RenaudWasTaken commented:
/lgtm

derekwaynecarr commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -115,6 +119,12 @@ type ContainerManager interface {
 	UpdateAllocatedDevices()
+
+	// GetAllDevices returns information about all the devices known to the manager
+	GetAllDevices() map[string]map[string]pluginapi.Device
```
we should update this for a more descriptive type response in the future than a map of maps to devices.

AlexeyPerevalov replied:
yes map[string]map[string]pluginapi.Device isn't so readable as e.g. DevicesPerResourcesMap or just DeviceResourceMap or ResourceDeviceMap.
that time it was release pressure, but now I think we can think about more readable name.

derekwaynecarr commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -77,6 +77,13 @@ type Manager interface {
+	// cpus for the container
+	GetCPUs(podUID, containerName string) []int64
+
+	// GetAllCPUs provides the ordered IDs of all the online CPUs found on the node.
```
its not really all right?  its just the list of cpus that are schedulable for the node, right?

derekwaynecarr replied:
would expect this to match language here:

	// GetAssignableCPUs returns the assignable (not allocated) CPUs

fromanirh replied:
yes, stale comment, like you said. Will update. Thanks for catching this!

derekwaynecarr commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -945,8 +949,6 @@ func (m *ManagerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Co
 			return nil, err
 		}
 	}
-	m.mutex.Lock()
```
just recording we appear to move this to an RLock inside poddevices.

fromanirh replied:
ack, missed in my rebase. Needs to be checked.

derekwaynecarr commented on test/e2e_node/podresources_test.go:
```diff
@@ -0,0 +1,380 @@
+	podResources, noResources = getPodResources(cli)
+	framework.ExpectEqual(len(podResources), 2)
+	framework.ExpectEqual(len(noResources), 2+1) // test pods + sriovdp
+	// TODO check for specific pods
```
am I missing something?  there is nothing that appears to verify that cpus are actually appearing in the response?

fromanirh replied:
right, missing (lost in rebases). Will take care.

derekwaynecarr commented on test/e2e_node/podresources_test.go:
```diff
@@ -0,0 +1,380 @@
+	devs := resp.GetDevices()
+	cpuIDs := resp.GetCpuIds()
+
+	onlineCPUIds, err := getOnlineCPUs()
```
i see this covers the all assignable cpus case, but nothing appears to check the actual cpu assignment for a pod is popuated.  is that intended to come later?

fromanirh replied:
yes, but since I need to work to address the other comments I'll try to squeeze it in.

derekwaynecarr reviewed (changes requested):
I have a few questions on intended scope of node e2e.

fromanirh commented:
/retest

fromanirh commented:
now totally hitting https://github.com/kubernetes/test-infra/issues/19910 . Retrying.

fromanirh commented:
/retest

fromanirh commented:
/retest
reason: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/95734/pull-kubernetes-bazel-build/1326913792403050496/build-log.txt
-> merge: ec3f3c5dbfc951f3d84fd19cfb91a75cb62bac5e - not something we can merge

fromanirh commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -276,6 +277,10 @@ func TestAllocatableResources(t *testing.T) {
 			}
+
+			sort.Slice(resp.CpuIds, func(i, j int) bool { return resp.CpuIds[i] < resp.CpuIds[j] })
+			sort.Slice(resp.Devices, func(i, j int) bool { return resp.Devices[i].ResourceName < resp.Devices[j].ResourceName })
```
not good enough. This has still room for false negatives. Needs another update.

swatisehgal replied:
Yes, as we are just sorting by resource name in case of devices if the resource name is same it we can't be sure of the order in which we have numa nodes  which could lead to mismatch because of order
eg example.com/deviceA on numa node 0 and example.com/deviceA on numa node 1 we can't sure on the way they are sorted and hence could still return a false negative when we compare the expected and obtained response strings.

swatisehgal commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -276,9 +278,57 @@ func TestAllocatableResources(t *testing.T) {
 			}
-			if tc.expectedAllocatableResourcesResponse.String() != resp.String() {
+
+			if !equalAllocatableResourcesResponse(tc.expectedAllocatableResourcesResponse, resp) {
```
Much better now :)

swatisehgal commented:
/retest

AlexeyPerevalov commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -1056,6 +1056,21 @@ func (m *ManagerImpl) isDevicePluginResource(resource string) bool {
 }
 
+// GetAllDevices returns information about all the devices known to the manager
+func (m *ManagerImpl) GetAllDevices() map[string]map[string]pluginapi.Device {
```
If here will be not a  pluginapi.Device, but appropriate version of podresourcesapi.ContainerDevices, like in GetDevices, I think, you could avoid dependencies in cm and podresources packages. I don't see extra field in Device, which doesn't exists in ContainerDevices.

swatisehgal replied:
Good point Alexey! Don't see it as a major blocker for the feature though, will address this in a follow-up PR.

kikisdeliveryservice commented:
As per https://github.com/kubernetes/enhancements/issues/2043#issuecomment-726545916 Code Freeze is now in effect, please file an Exception Request if you believe this needs to be in the 1.20 Release.

Best,
Kirsten

/milestone clear

fromanirh commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

fromanirh commented:
let's check if this PR has the same issue fixed in https://github.com/kubernetes/kubernetes/pull/96553

fromanirh commented:
job failed but apparently for different reasons:
` I1113 09:51:34.986] Nov 13 09:51:30.915: FAIL: Timed out after 600.000s. `
retrying.

fromanirh commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

fromanirh commented:
/retest

fromanirh commented:
Lane failed for deployment failure:
```
 I1113 10:28:40.686] Nov 13 10:28:35.544: INFO: Nvidia GPUs not available on Node: "e2e-0cbc78e434-8baae-minion-group-1x2w"
I1113 10:28:40.687] Nov 13 10:28:36.327: INFO: Get container nvidia-gpu-device-plugin-4vvvj/nvidia-gpu-device-plugin usage on node e2e-0cbc78e434-8baae-minion-group-1x2w. CPUUsageInCores: 5.0195e-05, MemoryUsageInBytes: 9080832, MemoryWorkingSetInBytes: 2916352
I1113 10:28:40.687] Nov 13 10:28:36.327: INFO: Get container nvidia-driver-installer-fp9zz/nvidia-driver-installer usage on node e2e-0cbc78e434-8baae-minion-group-1x2w. CPUUsageInCores: 1.128646821, MemoryUsageInBytes: 4093575168, MemoryWorkingSetInBytes: 655339520
I1113 10:28:40.687] Nov 13 10:28:36.425: INFO: Get container nvidia-driver-installer-ggdls/nvidia-driver-installer usage on node e2e-0cbc78e434-8baae-minion-group-mfq5. CPUUsageInCores: 1.152389565, MemoryUsageInBytes: 4040839168, MemoryWorkingSetInBytes: 655065088
I1113 10:28:40.687] Nov 13 10:28:36.425: INFO: Get container nvidia-gpu-device-plugin-mp2cg/nvidia-gpu-device-plugin usage on node e2e-0cbc78e434-8baae-minion-group-mfq5. CPUUsageInCores: 2.8394e-05, MemoryUsageInBytes: 9080832, MemoryWorkingSetInBytes: 2748416
I1113 10:28:40.687] Nov 13 10:28:36.544: INFO: Getting list of Nodes from API server
I1113 10:28:40.688] Nov 13 10:28:36.667: INFO: gpuResourceName nvidia.com/gpu
I1113 10:28:40.688] Nov 13 10:28:36.667: INFO: Nvidia GPUs not available on Node: "e2e-0cbc78e434-8baae-minion-group-1x2w"
I1113 10:28:40.688] Nov 13 10:28:36.667: FAIL: Timed out after 600.092s. 
```

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

RenaudWasTaken commented on test/e2e_node/podresources_test.go:
```diff
@@ -279,9 +306,62 @@ var _ = SIGDescribe("POD Resources [Serial] [Feature:PODResources][NodeFeature:P
+
+		ginkgo.It("should return the expected device count", func() {
+			// this is a very rough check. We just want to rule out system that does NOT have any SRIOV device.
+			if sriovdevCount, err := countSRIOVDevices(); err != nil || sriovdevCount == 0 {
```
Do we have any machine in the test infrastructure that has SRIOV devices? If so how and when are tests ran, if not what is our plan for continuous integration?

fromanirh replied:
at this moment, no. This is a problem for few other e2e tests. There is an ongoing conversation - which I need to push forward more - about if and how we can add these lanes to the k8s ci.
I have few internal jenkins instances inside RH on which we continuously run the tests. Another option is somehow hooking these instances to k8s prow, but it's challenging as well.

The problem is to get any device supported by device plugins, sriov happens to be the easiest for me to get access to but I'm very open to use any other device we can use.
The only requirement is to be a real device, no fakes/mocks (maybe we can re-evaluate this requirement?)

RenaudWasTaken commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -391,6 +391,11 @@ func runTopologyManagerPolicySuiteTests(f *framework.Framework) {
 	runMultipleGuPods(f)
 }
 
+// waitForAllContainerRemoval waits until all the containers on a given pod are really gone.
```
Nit on commit message:

```
document why teardowbnSRIOVPod was to wait for all the containers
to be gone before to end, and why is important.
```

should be:
```
document why teardownSRIOVPod has to wait for all the containers
to be gone before to end, and why is important.
```

fromanirh replied:
ack, will fix since I want to address the last pending comment

fromanirh replied:
actually not sure because I don't want to lose the LGTM if I can help it :\

RenaudWasTaken reviewed (approved):
/lgtm

Can you address @derekwaynecarr 's comment about `map[string]map[string]pluginapi.Device` either in a followup PR or this PR :) ?

Thanks!

AlexeyPerevalov commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -70,3 +70,33 @@ func (p *v1PodResourcesServer) List(ctx context.Context, req *v1.ListPodResource
+
+	for resourceName, resourceDevs := range allDevices {
+		for devID, dev := range resourceDevs {
+			for _, node := range dev.GetTopology().GetNodes() {
```
func (m* Device) GetTopology gould return nil, if device plugin doesn't fill it, as we know now some device plugin in case of non NUMA architecture .is doing it

fromanirh replied:
Thanks for the reminder! Using the Get*() accessor makes us safe from this case, and I added this https://github.com/kubernetes/kubernetes/pull/95734/commits/cd06adb00b30bf7dd9ce115b3c07b112b94ec735 to test this occurrence.

AlexeyPerevalov replied:
But if we don't have topology, but have devices, do we still need to return it? For our use case not, but maybe someone else doesn't require topology... hmm in this case it's better to rely on v1.Node...Allocatable.

fromanirh replied:
True. Let me review what List() does in this case, we need to be consistent (or to clearly document why we are not consistent...)

fromanirh replied:
Decided to report devices without topology information, let the client decide.

mythi commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -18,6 +18,15 @@ option (gogoproto.goproto_unrecognized_all) = false;
+
+message AllocatableResourcesRequest {}
+
+// AllocatableResourcesResponses contains informations about all the devices known by the kubelet
+message AllocatableResourcesResponse {
+    repeated ContainerDevices devices = 1;
+    repeated int64 cpu_ids = 2;
```
for the sake of completeness, can this should be made to include *all* container resources. There are use cases where, e.g., extended resources add value to get the big picture so be able to monitor full set of container resources is important.

fromanirh replied:
AFAIU this is an implementation detail and don't requires API (or KEP) changes, right?

mythi replied:
`AllocatableResourceResponse`  likely needs changes but it's in this PR so changes are possible.

To give an idea about the gap: I have a device plugin (`sgx.intel.com/enclave`) to allow containers to use devices nodes (`/dev/sgx_enclave`) and a separate extended resource to track how much SGX memory is needed (`sgx.intel.com/epc`).

With this API, I'm currently not able to monitor how much SGX memory each container is using. I can only see who are using the device node (which carries no resource information a such).

fromanirh replied:
Thanks for presenting this use case. 

This use case seems more a possible a gap in the List() endpoint rather in the `GetAllocatableResources` endpoint (which this PR concerns to).
In `GetAllocatableResources` we want to report how many resources does the node have, not the details of their usage (both in the sense of which pod uses them and the details of each resource usage, in this example the memory amount).
It still seems to me we need no changes in the `api,proto` spec. We can implement the reporting of the external resources, possibly in another PR.

In any case, should we need an `api.proto` extension, we will need to amend the KEP, which calls for more, separate PRs.

fromanirh replied:
KEP describing the original podresources API: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/606-compute-device-assignment
KEP extension this PR is (partially) implementing: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2043-pod-resource-concrete-assigments

mythi replied:
> In any case, should we need an api.proto extension, we will need to amend the KEP, which calls for more, separate PRs.

@frobware thanks for sharing the KEPs! I had missed the original one. I'll need to take a closer look which one (or is it both) need updates

fromanirh commented:
The last upload address missing comments from @AlexeyPerevalov , and suggestions @RenaudWasTaken and @derekwaynecarr added previously.

fromanirh commented:
/retest

kad commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -70,3 +70,45 @@ func (p *v1PodResourcesServer) List(ctx context.Context, req *v1.ListPodResource
+					DeviceIds:    []string{devID},
+					Topology: &v1.TopologyInfo{
+						Nodes: []*v1.NUMANode{
+							{ID: node.GetID()},
```
it can be multiple ids per device, as announced by device plugin.

fromanirh replied:
ack, I need to fix this.

kad commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -1060,6 +1060,40 @@ func (m *ManagerImpl) isDevicePluginResource(resource string) bool {
+						// note: each pluginapi.Device has a single ID but can have
+						// a slice of NUMANodes on its TopologyInfo. Since the same ID
+						// cannot be on multiple NUMA Node, we just pick the first value.
+						{ID: nodes[0].GetID()},
```
that is wrong assumption. One device ID can have multiple "numa" node ids.
Device is connected to PCI bus. PCI bus is a property of CPU die (or e.g. I/O die in AMD Rome). this die can span multiple "numa" nodes, as NUMA id represent only memory controller number and not guarantee 1-to-1 map to e.g. PCI bus.

fromanirh replied:
Thanks for pointing this out. This indeed voids the assumption. So we can have (and thus we should report) the same device ID in multiple NUMA nodes, right?

kad replied:
better not to use numa id as a key at all. ideally it should be presented in "zone" level as we discussed in scheduler extension crds, but as we don't have those now, for this particular API if we want to express internal representation of TopologyManager that using memory controller numbers as "zone id", then yes, announce same device multiple times for each ID, according to topology hint received from device plugin.

fromanirh replied:
ack, thanks for pointing this out. This indeed something I for myself need to fully absorb (I'm used to simpler topologies) so thanks for taking the time to explain it again.
For the purposes of this PR, indeed we reuse the same internal representation Topology Manager uses, but I'll strive to take your points into account. I'll post an update soon.

fromanirh replied:
done in https://github.com/kubernetes/kubernetes/pull/95734/commits/521ee88a0f3142650eb235e8ea93b8e6e825f773

ffromani committed:
podresources: devices: add test for dev reporting

Add test to reflect the correct behaviour according to
review comments.

Most notably, we should consider that -as the device plugin API
allows to express- a device ID can have multiple "NUMA" node IDs.
(example: AMD Rome).
More details:
https://github.com/kubernetes/kubernetes/pull/95734#discussion_r539545041

Signed-off-by: Francesco Romani <fromani@redhat.com>

AlexeyPerevalov reviewed (approved):
/lgtm

derekwaynecarr commented:
/milestone v1.21

fromanirh commented:
will rebase to handle the merge conflict ASAP.

ffromani committed:
node: podresources: add generated files

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/retest

fromanirh commented:
/retest

ffromani committed:
podresources: getallocatable: add feature gate

Add feature gate to disable the GetAllocatableResources API.
The feature gate isd alpha stage, disabled by default.

Add e2e test to demonstrate the behaviour with feature gate disabled.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/retest

fromanirh commented:
/retest

ffromani committed:
kubelet: podresources: per-endpoint metrics

Before the addition of GetAllocatableResources, the
podresources API had just one endpoint `List()`, thus we could just
account for the total of the calls to have a good pulse of the API usage.
Now that we extend the API with more endpoints
(`GetAlloctableResources`), in order to improve the observability we add
per-endpoint counters, in addition to the existing counter of the total
API calls.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
Doesn't seem to be caused by this PR:
```
Kubernetes e2e suite: [sig-cli] Kubectl client Simple pod should return command exit codes expand_more               1m39s | Kubernetes e2e suite: [sig-cli] Kubectl client Simple pod should return command exit codes expand_more | 1m39s
-- | -- | --
Kubernetes e2e suite: [sig-cli] Kubectl client Simple pod should return command exit codes expand_more | 1m39s
Kubernetes  e2e suite: [sig-storage] PersistentVolumes-local  [Volume type:  dir-link] Two pods mounting a local volume at the same time should be  able to write from pod1 and read from pod2 expand_more | Kubernetes  e2e suite: [sig-storage] PersistentVolumes-local  [Volume type:  dir-link] Two pods mounting a local volume at the same time should be  able to write from pod1 and read from pod2 expand_more
Kubernetes  e2e suite: [sig-storage] PersistentVolumes-local  [Volume type:  dir-link] Two pods mounting a local volume at the same time should be  able to write from pod1 and read from pod2 expand_more
```

fromanirh commented:
/retest

tengqm commented on pkg/features/kube_features.go:
```diff
@@ -681,6 +681,12 @@ const (
+	// owner: @fromani
+	// alpha: v1.21
+	//
+	// Enable POD resources API to return allocatable compute resources
```
This "compute resources" is a little bit misleading. Please consider just calling it "allocatable resources".

fromanirh replied:
ok, renaming to "allocatable resources"

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

derekwaynecarr commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -70,3 +74,46 @@ func (p *v1PodResourcesServer) List(ctx context.Context, req *v1.ListPodResource
+
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletPodResourcesGetAllocatable) {
+		metrics.PodResourcesEndpointErrorsGetAllocatableCount.WithLabelValues("v1").Inc()
+		return nil, fmt.Errorf("POD Resources API GetAllocatableResources disabled")
```
nit: "POD" -> "Pod"

fromanirh replied:
ack

derekwaynecarr commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -118,6 +118,12 @@ type ContainerManager interface {
+	GetAllDevices() podresourcesapi.ResourceDeviceMap
+
+	// GetAllCPUs returns the assignable (not allocated) CPUs
+	GetAllCPUs() []int64
```
should we just call this `GetAllocatableCPUs`?

fromanirh replied:
I see your point. The original idea was that from the manager perspective the API is returning all the resources - CPUs in this case. But being consistent all across the stack is probably better, so I'll change accordingly.

derekwaynecarr commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -118,6 +118,12 @@ type ContainerManager interface {
 	UpdateAllocatedDevices()
+
+	// GetAllDevices returns information about all the devices known to the manager
+	GetAllDevices() podresourcesapi.ResourceDeviceMap
```
should we call this `GetAllocatableDevices` ?

fromanirh replied:
sure, same original rationale as per `GetAllCPUs` above. Changing.

derekwaynecarr reviewed (changes requested):
Can we use consistent terminology?

We have phrasing for Allocatable/All/Assignable that is confusing throughout.

It appeared to me (unless I am mistaken) that we could have used Allocatable throughout?

Let me know if I misunderstood.

derekwaynecarr commented:
/assign @klueska 

can you assist in looking at changes in corresponding resource managers as well?

derekwaynecarr commented:
/assign

fromanirh commented:
Will probably need to rebase on top of https://github.com/kubernetes/kubernetes/pull/99634 - monitoring.

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
rebased to fix the conflict

klueska commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -18,6 +18,15 @@ option (gogoproto.goproto_unrecognized_all) = false;
+message AllocatableResourcesRequest {}
+
+// AllocatableResourcesResponses contains informations about all the devices known by the kubelet
+message AllocatableResourcesResponse {
+    repeated ContainerDevices devices = 1;
+    repeated int64 cpu_ids = 2;
```
Why not just have this return a `repeated ContainerResources` instead of duplicating the fields of this in here?

fromanirh replied:
I need to doublecheck the revision history, but I think we can change as you suggested, which seems nicer indeed.

swatisehgal replied:
`AllocatableResourcesResponse`  should return cumulative data of all the devices and cpus known to kubelet i.e. data corresponding to not only resources allocated to containers but also the ones that haven't been allocated. Therefore, we should keep it the way we have it now instead of using `repeated ContainerResources` because changing to `repeated ContainerResources` would mean we would only account for resources that have been allocated to containers.

fromanirh replied:
Thanks @swatisehgal! It boils down to how resources are accounted indeed and how we report to clients; yes, there are similarities, but reusing the `ContainerResources`, albeit doable (the fields are indeed similar) will make harder for the clients to consume the data. Therefore I'm keeping the current API.

klueska commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -118,6 +119,12 @@ type ContainerManager interface {
+	GetAllocatableDevices() map[string]map[string]pluginapi.Device
+
+	// GetAllocatableCPUs returns the assignable (not allocated) CPUs
+	GetAllocatableCPUs() []int64
```
Is there a reason not to have this return a `cpuset.CpuSet()`?  Using tzhis tye makes it more flexible if we need to perform any operations on the returned CPUs in the future.

Likewise, I saw a discussion on wrapping `map[string]map[string]pluginapi.Device` in its own type (which I agree with), which would then make this symmetzric with that.

fromanirh replied:
There is no good reason that I can recall, so I'll change accordingly.
Same goes for wrapping `map[string]map[string]pluginapi.Device` in a new type, which I believe can sit inside the devicemanager package, right?

klueska replied:
Yes. I would put it inside `pod_devices.go`.

fromanirh replied:
Perfect. On it.

fromanirh commented:
e2e-azure-disk-windows seems to fail much more often in certain times of the day, I wonder if it is related to the load? Not even sure this qualifies as a flake

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -85,6 +85,9 @@ type Manager interface {
 	GetPodTopologyHints(pod *v1.Pod) map[string][]topologymanager.TopologyHint
+
+	// GetAllocatableCPUs returns the assignable (not allocated) CPUs
+	GetAllocatableCPUs() []int64
```
I think this should return a `cpuset.CPUSet` just like the call to `GetCPUs()` above.
Also, can you move this call just below `GetCPUs()` to group them together better.

fromanirh replied:
ack

fromanirh replied:
Do we want, in a future, lower priority PR, change the return type of `GetCPUs` so it returns `cpuset.CPUSet`, to make the code consistent and improve the flexibility?

klueska replied:
If it's not too much effort, I think it would be good to do it now.

fromanirh replied:
Fair enough. I'll see what I can do (in a separate commit)

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -124,6 +127,9 @@ type manager struct {
 	stateFileDirectory string
+
+	// assignableCPUs is the set of online CPUs as reported by the system
+	assignableCPUs cpuset.CPUSet
```
Is there a reason not to call this variable `allocatableCPUs` to make it match the API call now that we've settled on the name `Allocatable` for the API call?

fromanirh replied:
The only reason is to minimize the change internal to the module, which is a debatable reason (see below) and which I'm fine to change.

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -296,6 +305,18 @@ func (m *manager) GetPodTopologyHints(pod *v1.Pod) map[string][]topologymanager.
 }
 
+func (m *manager) GetAllocatableCPUs() []int64 {
+	// please note that 'assignableCPUs' is the old internal naming which predated
+	// the introduction of the `GetAllocatableCPUs` API.
+	// We keep the old name internally for consistency within the cpumanager code.
```
I don't think this is necessary -- let's just rename it. Also, with my suggestion above all of the code below this goes away, and we just return `m.allocatableCPUs`.

fromanirh replied:
agreed.

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager_test.go:
```diff
@@ -120,6 +120,10 @@ func (p *mockPolicy) GetPodTopologyHints(s state.State, pod *v1.Pod) map[string]
 	return nil
 }
 
+func (p *mockPolicy) GetAssignableCPUs(m state.State) cpuset.CPUSet {
+	return cpuset.NewCPUSet()
+}
+
```
Shouldn't this be `GetAllocatableCPUs()`?

fromanirh replied:
yes, forgot to update/lost in rebase. Will update.

klueska commented on pkg/kubelet/cm/cpumanager/policy.go:
```diff
@@ -38,4 +39,6 @@ type Policy interface {
 	// among this and other resource controllers.
 	GetPodTopologyHints(s state.State, pod *v1.Pod) map[string][]topologymanager.TopologyHint
+	// GetAssignableCPUs returns the assignable (not allocated) CPUs
+	GetAssignableCPUs(m state.State) cpuset.CPUSet
```
Let's call this `GetAllocatableCPUs()` for consistency.

fromanirh replied:
ack

klueska commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -1068,6 +1068,21 @@ func (m *ManagerImpl) isDevicePluginResource(resource string) bool {
+			resp[resourceName][devID] = dev
+		}
+	}
+	klog.V(4).Infof("known devices: %d", len(resp))
```
Is this just making a copy of `allDevices`?
If we put a type around  `map[string]map[string]pluginapi.Device` we can just create a `Clone()` call on it to simplify this here.

fromanirh replied:
good point, changing as suggested

klueska commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -118,6 +118,12 @@ type ContainerManager interface {
 	UpdateAllocatedDevices()
+
+	// GetAllocatableDevices returns information about all the devices known to the manager
+	GetAllocatableDevices() podresourcesapi.ResourceDeviceMap
+
+	// GetAllocatableCPUs returns the assignable (not allocated) CPUs
+	GetAllocatableCPUs() []int64
```
It feels a bit strange to have the `GetAllocatableDevices()` call return something very specific to the `podresourcesapi`. I would have expected a more generic type that then get's translated into the `podresourcesapi.ResourceDeviceMap` at a higher layer.

fromanirh replied:
good point, let me fix

fromanirh replied:
We introduced `podresourcesapi.ResourceDeviceMap` to abstract the type and avoid the awkward `map[string]map[string]pluginapi.Device` and to possibly avoid using the `pluginapi` dependency in the upper layers.
Moving the `map[string]map[string]pluginapi.Device` inside `devicemanager` is similar and probably a bit better because indeed we don't have a specific dependency on `podresourcesapi`.
So, if we go this direction, it makes little to none sense to have an intermediate step translating to `ResourceDeviceMap`. We can do just one translation to the public API value in the `podresources` server. 
Hence I removed this commit and simplified the series.

klueska reviewed (changes requested):
I did a quick review of all of the changes, and a more thorough review of the changes under `pkg/kubelet/cm`. I will be available to iterate quickly for the next 4 (working) days until the code freeze to address any remaining comments.

ffromani committed:
node: podresources: make GetCPUs return cpuset

a upcoming patch wants to add GetAllocatableCPUs() returning a cpuset.
To make the code consistent and a bit more flexible, we change the
existing interface to also return a cpuset.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: podresources: list devices without topology

It's legal for device plugins to not expose topology informations.
Previously, the code was just skipping these devices.
Review highlighted is better to report them anyway and let the
client application decide if they still want somehow to track them
or skip them entirely.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
podresources: test: add test for nil Topology

From https://github.com/kubernetes/kubernetes/pull/96553
we are reminded we need to handle the case on which
a device plugin reports nil Topology, which is legal.
Add unit test to ensure this case is handled.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
Rebased to fix another conflict.
Addressed all comments, please note the `api.proto` was not changed (explained the reasoning in inline comments)
Squashed together few commits where made sense - commits were separated because of historical reasons and/or because spawned to answer review comments.

fromanirh commented:
fixed and further improved unit tests. Now ready for review again.

klueska commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -367,11 +367,9 @@ func (p *staticPolicy) GetTopologyHints(s state.State, pod *v1.Pod, container *v
-	// Get a list of available CPUs.
-	available := p.assignableCPUs(s)
-
 	// Get a list of reusable CPUs (e.g. CPUs reused from initContainers).
 	// It should be an empty CPUSet for a newly created pod.
+	available := p.GetAllocatableCPUs(s)
 	reusable := p.cpusToReuse[string(pod.UID)]
```
nit: why was this change made?

fromanirh replied:
Good catch. It's an editing artifact, no reason. Will revert

klueska commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -346,3 +346,21 @@ func (pdev *podDevices) getContainerDevices(podUID, contName string) []*podresou
 }
+
+// ResourceDeviceInstances is a map ping resource name -> device name -> device data
+type ResourceDeviceInstances map[string]map[string]pluginapi.Device
```
I think we can come up with a better name for this (though I don't have any suggestions off the top of my head). Possibly something with the name `plugin` in it instead of `resource` to make it clear that this holds `pluginapi.Devices`. Maybe even just `PluginDevices`. Or maybe just `DeviceSet` to mirror `CPUSet`. Not sure.

Let's do that in a follow-up PR though.

fromanirh replied:
Yes, not fond of the name either. `DeviceSet` is tempting, but I think the name should convey there are nested maps. Will think about this.

klueska commented on pkg/kubelet/apis/podresources/types.go:
```diff
@@ -19,12 +19,15 @@ package podresources
 type DevicesProvider interface {
 	GetDevices(podUID, containerName string) []*podresourcesapi.ContainerDevices
 	UpdateAllocatedDevices()
+	GetAllocatableDevices() devicemanager.ResourceDeviceInstances
```
Hmm. I think you propagated this a bit too far. Here is where I would expect this now to be of type `podresourcesapi.ContainerDevices`. It is at this layer that the translation from `devicemanager.ResourceDeviceInstances` to ``podresourcesapi.ContainerDevices` should happen.

fromanirh replied:
This is a good point. The problem is - and here's why I didn't implement the change yet - is this layer is up until now extremely thin, like interface thin. We pass to the podresources server the concrete managers that implement the interface, without middleman translation layer. See: https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/kubelet.go#L2237 , where `containerManager` here is a https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/cm/container_manager.go#L44 .
So it takes a bit of rearchitecting. Let me think about it.

klueska replied:
OK, well maybe the right thing to do then, is to explicitly make the `containermanager` a `podresources.DevicesProvider` and a `podresources.CPUsProvider`. 

Then it wouldn't be strange for it to be returning these types.

klueska replied:
i.e., make the `contaienrmanager` interface:
```
type ContainerManager interface {
	// Runs the container manager's housekeeping.
	// - Ensures that the Docker daemon is in a container.
	// - Creates the system container where all non-containerized processes run.
	Start(*v1.Node, ActivePodsFunc, config.SourcesReady, status.PodStatusProvider, internalapi.RuntimeService) error
	// SystemCgroupsLimit returns resources allocated to system cgroups in the machine.
	// These cgroups include the system and Kubernetes services.
	SystemCgroupsLimit() v1.ResourceList
	// GetNodeConfig returns a NodeConfig that is being used by the container manager.
	GetNodeConfig() NodeConfig
	// Status returns internal Status.
	Status() Status
	// NewPodContainerManager is a factory method which returns a podContainerManager object
	// Returns a noop implementation if qos cgroup hierarchy is not enabled
	NewPodContainerManager() PodContainerManager
	// GetMountedSubsystems returns the mounted cgroup subsystems on the node
	GetMountedSubsystems() *CgroupSubsystems
	// GetQOSContainersInfo returns the names of top level QoS containers
	GetQOSContainersInfo() QOSContainersInfo
	// GetNodeAllocatableReservation returns the amount of compute resources that have to be reserved from scheduling.
	GetNodeAllocatableReservation() v1.ResourceList
	// GetCapacity returns the amount of compute resources tracked by container manager available on the node.
	GetCapacity() v1.ResourceList
	// GetDevicePluginResourceCapacity returns the node capacity (amount of total device plugin resources),
	// node allocatable (amount of total healthy resources reported by device plugin),
	// and inactive device plugin resources previously registered on the node.
	GetDevicePluginResourceCapacity() (v1.ResourceList, v1.ResourceList, []string)
	// UpdateQOSCgroups performs housekeeping updates to ensure that the top
	// level QoS containers have their desired state in a thread-safe way
	UpdateQOSCgroups() error
	// GetResources returns RunContainerOptions with devices, mounts, and env fields populated for
	// extended resources required by container.
	GetResources(pod *v1.Pod, container *v1.Container) (*kubecontainer.RunContainerOptions, error)
	// UpdatePluginResources calls Allocate of device plugin handler for potential
	// requests for device plugin resources, and returns an error if fails.
	// Otherwise, it updates allocatableResource in nodeInfo if necessary,
	// to make sure it is at least equal to the pod's requested capacity for
	// any registered device plugin resource
	UpdatePluginResources(*schedulerframework.NodeInfo, *lifecycle.PodAdmitAttributes) error
	InternalContainerLifecycle() InternalContainerLifecycle
	// GetPodCgroupRoot returns the cgroup which contains all pods.
	GetPodCgroupRoot() string
	// GetPluginRegistrationHandler returns a plugin registration handler
	// The pluginwatcher's Handlers allow to have a single module for handling
	// registration.
	GetPluginRegistrationHandler() cache.PluginHandler
	// ShouldResetExtendedResourceCapacity returns whether or not the extended resources should be zeroed,
	// due to node recreation.
	ShouldResetExtendedResourceCapacity() bool
	// GetAllocateResourcesPodAdmitHandler returns an instance of a PodAdmitHandler responsible for allocating pod resources.
	GetAllocateResourcesPodAdmitHandler() lifecycle.PodAdmitHandler
	// UpdateAllocatedDevices frees any Devices that are bound to terminated pods.
	UpdateAllocatedDevices()
        // Implements the podresources Provider API for CPUs and Devices
	podresources.CPUsProvider
	podresources.DevicesProvider
```

klueska replied:
Which is actually very similar to what you had originally (sorry for the churn), but now it is much more explicit.

fromanirh replied:
Nicer indeed! At first it felt to me a little weird to push `UpdateAllocatedDevices` into the `podresources.DevicesProvider` interface, but on second thought it's OK: the only other big consumer is `devicemanager` itself, and `ContainerManager` is just exposing this API to enable the `podresources` server. So it's now clearer that `ContainerManager` is just a delegating shim here.

klueska commented on pkg/kubelet/apis/podresources/types.go:
```diff
@@ -34,5 +37,6 @@ type PodsProvider interface {
 
 // CPUsProvider knows how to provide the cpus used by the given container
 type CPUsProvider interface {
-	GetCPUs(podUID, containerName string) []int64
+	GetCPUs(podUID, containerName string) cpuset.CPUSet
+	GetAllocatableCPUs() cpuset.CPUSet
```
Likewise here -- this is where I would have expected it to be translated from a `CPUSet` to whatever type the `podresourcesapi` requires.

fromanirh replied:
ack, same as above.

klueska commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -107,7 +108,10 @@ type ContainerManager interface {
 	GetDevices(podUID, containerName string) []*podresourcesapi.ContainerDevices
```
I know this isnÄt part of your change -- but for consistency, this really should be `devicemanager.ResourceDeviceInstances`  as it now is for `GetAllocatableDevices()` below.

fromanirh replied:
fair point.

ffromani committed:
node: podresources: make GetDevices() consistent

We want to make the return type of the GetDevices() method of the
podresources DevicesProvider interface consistent with
the newly added GetAllocatableDevices type.
This makes the code easier to read and reduces the coupling between
the podresourcesapi server and the devicemanager code.

No intended changes in behaviour, but the different return types
now requires some data massaging. Tests are updated accordingly.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: podresources: translate types in cm

during the review, we convened that the manager types
(CPUSet, ResourceDeviceInstances) should not cross the
containermanager API boundary; thus, the ContainerManager layer
is the correct place to do the type conversion

We push back the type conversions from the podresources server
layer, fixing tests accordingly.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/retest

fromanirh commented:
I think the lane failures are flakes, hence I consider this new version ready for review again (cc @klueska - thanks for the great reviews and discussion so far!)

Since last upload, besides another needed rebase, I pushed back the type translation in the ContainerManager because, after review, it felt a more natural place. The bulk of the changes happen in the new commit https://github.com/kubernetes/kubernetes/pull/95734/commits/670ddd58918f573d9b1a471367fce2758c0b87e9 - other commits after that have minor changes to deal with the internal API changes.

Another option would have been to make `cpumanager` and `devicemanager` implementors of the `{CPU,Device}Provider` interfaces that the `podresources` requires. This however would have required more refactoring to break dep cycles, while the current approach mostly moves code around.

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/test pull-kubernetes-e2e-azure-file-windows-containerd

klueska commented on pkg/kubelet/cm/cpuset/cpuset.go:
```diff
@@ -75,6 +75,15 @@ func NewCPUSet(cpus ...int) CPUSet {
 }
 
+// NewCPUSet returns a new CPUSet containing the supplied elements, as slice of int64.
+func NewCPUSetInt64(cpus ...int64) CPUSet {
+	b := NewBuilder()
+	for _, c := range cpus {
+		b.Add(int(c))
+	}
+	return b.Result()
+}
+
```
Is this used anywhere?

fromanirh replied:
yes: https://github.com/kubernetes/kubernetes/pull/95734/commits/0e5f114a1ca678d7b0a406647766e141e4147f88#diff-5cb7ca1bf56d1b07b77b5bd67b9be08c8bf81063529fc6e32fb9521cbb15b6c2R444

fromanirh replied:
Do you want me to move this code in `podresources_test.go` ?

klueska replied:
No. It’s fine here and could prove useful for others

fromanirh replied:
ack

klueska commented on test/e2e_node/podresources_test.go:
```diff
@@ -0,0 +1,754 @@
+
+	if reservedSystemCPUs.Size() > 0 {
+		cpus := reservedSystemCPUs.String()
+		framework.Logf("configureCPUManagerInKubelet: using reservedSystemCPUs=%q", cpus)
```
is this log message wrong? Since this function is called `configurePODResourcesInKubelet`.

fromanirh replied:
yep, this is a copy/paste relic. Fixing.

klueska reviewed (commented):
Left a few small comments, but nothing blocking. Thanks for working through the review with me, I think it's much cleaner now.


We can look into making the `CPUManager` and the `devicemanager` an explicit `podresources.CPUProvider` and `podresourcces.DeviceProvider` in a future PR.

fromanirh commented:
It seems I'm hitting

fromanirh commented:
it seems I'm hitting https://github.com/kubernetes/kubernetes/issues/98084

klueska commented:
/lgtm
/approve

For kubelet changes

fromanirh commented:
/retest

klueska commented:
/cc @feature-approvers 
For final approval

klueska commented:
/cc @derekwaynecarr 
For final approval

fromanirh commented:
The was `pull-kubernetes-e2e-gce-device-plugin-gpu` lane was  triggered explicitely (not required by pre-merge tide checks) to validate https://github.com/kubernetes/kubernetes/pull/95734/commits/23171f142a4cd15d882b773f16f4b8f710c08d70 - but it seems it was flaky.

fromanirh commented:
/retest

erismaster commented:
Hi 👋 Bug Triage here, just a reminder that code freeze is end of day, tomorrow the 9th! PRs need to be merged in before then, or they will require an approved exception to go into the 1.21 milestone.

fromanirh commented:
The azure-disk-windows lane seems broken: https://testgrid.k8s.io/sig-windows-azure#pr-k8s-azure-disk-e2e-master-windows

fromanirh commented:
also related

fromanirh commented:
filed: https://github.com/kubernetes/kubernetes/issues/99947

fromanirh commented:
/retest

klueska commented:
/lgtm

annajung commented:
Hi @fromanirh, a friendly reminder that the code freeze for 1.21 is today.
Please make sure to resolve the conflicting file and get the necessary reviews for this to merge by EOD PST.

fromanirh commented:
rebased to fix the conflict. No code changes besides the confict fix itself.

klueska commented:
/lgtm

fromanirh commented:
/test pull-kubernetes-e2e-kind

fromanirh commented:
/test  pull-kubernetes-integration

fromanirh commented:
/test pull-kubernetes-e2e-azure-disk-windows
/test pull-kubernetes-e2e-azure-file-windows

fromanirh commented:
/retest

derekwaynecarr reviewed (approved):
thank you for making the terminology updates.

/approve
/lgtm

dchen1107 commented:
/lgtm
/approve based on Kevin's review since I only reviewed API changes.
